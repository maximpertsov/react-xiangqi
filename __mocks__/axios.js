const mock = jest.genMockFromModule('axios');

export default mock;
