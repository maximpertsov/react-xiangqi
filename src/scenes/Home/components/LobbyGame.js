import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import { Button } from 'semantic-ui-react';

import client from 'services/client';
import { WebSocketContext } from 'services/WebSocketProvider';

const LobbyGame = ({ id, parameters }) => {
  const io = useContext(WebSocketContext);
  const username = useSelector(state => state.username);

  const joinGame = id => () => {
    client
      .patch(`game/request/${id}`, {
        player2: username,
      })
      .then(({ data: { game, player1, player2 } }) => {
        if (!game) return;

        io.send({
          type: 'joined_lobby_game',
          game,
          players: [player1, player2],
        });
      });
  };

  return (
    <Button className="LobbyGame" onClick={joinGame(id)} fluid>
      {`vs ${parameters.team || '?'}`}
    </Button>
  );
};

LobbyGame.propTypes = {
  id: PropTypes.number.isRequired,
  parameters: PropTypes.shape().isRequired,
};

export default LobbyGame;
