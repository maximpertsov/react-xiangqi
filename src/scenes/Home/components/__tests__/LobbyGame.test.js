import React, { useContext } from 'react';
import { useSelector } from 'react-redux';
import axios from 'axios';

import LobbyGame from '../LobbyGame';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn(),
}));
jest.mock('react-redux');
jest.mock('axios');

describe('LobbyGame', () => {
  const id = 123;
  const username = 'alice';
  const parameters = { team: 'red' };
  const gameSlug = 'abc123';
  const store = mockStore({ username });
  const io = { send: jest.fn() };

  let wrapper;

  beforeEach(() => {
    useContext.mockReturnValue(io);
    useSelector.mockImplementation(callback => callback(store.getState()));

    axios.patch.mockResolvedValue({
      data: {
        game: gameSlug,
        player1: 'bob',
        player2: username,
      },
    });

    wrapper = shallowWrappedComponent(
      <LobbyGame id={id} parameters={parameters} />,
      store,
    );
  });

  afterEach(() => {
    store.getActions();
    jest.resetAllMocks();
  });

  test('snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('click', async () => {
    expect.assertions(2);
    wrapper.find('Button').simulate('click');

    await expect(axios.patch).toHaveBeenCalledWith(`game/request/${id}`, {
      player2: username,
    });
    expect(io.send).toHaveBeenCalledWith({
      type: 'joined_lobby_game',
      game: gameSlug,
      players: ['bob', username],
    });
  });
});
