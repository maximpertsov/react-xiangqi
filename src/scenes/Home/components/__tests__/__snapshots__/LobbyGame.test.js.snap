// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LobbyGame snapshot 1`] = `
<Button
  as="button"
  className="LobbyGame"
  fluid={true}
  onClick={[Function]}
>
  vs red
</Button>
`;
